backend/server.js

```diff
@@ -1,29 +1,179 @@
+mongoose.connect(mongoUrl, {
+  useNewUrlParser: true,
+  useUnifiedTopology: true,
+  useCreateIndex: true
+});
+mongoose.Promise = Promise;
+
+const userSchema = new mongoose.Schema({
+  name: {
+    type: String,
+    required: true,
+    minlength: [2, 'Name is too short'],
+    maxlength: [40, 'Name is too long']
+  },
+  email: {
+    type: String,
+    unique: true,
+    required: true
+  },
+  password: {
+    type: String,
+    required: [true, 'Password is required'],
+    minlength: [5, 'Password must be at least 5 characters']
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex')
+  }
+});
+
+// Middleware to hash password before new user is saved
+userSchema.pre('save', async function (next) {
+  const user = this;
+
+  if (!user.isModified('password')) {
+    return next();
+  }
+  const salt = bcrypt.genSaltSync();
+  user.password = bcrypt.hashSync(user.password, salt);
+  next();
+});
+
+// Middleware to authenticate user
+const authenticateUser = async (req, res, next) => {
+  try {
+    const accessToken = req.header('Authorization');
+    const user = await User.findOne({ accessToken });
+    if (!user) {
+      throw 'User not found';
+    }
+    req.user = user;
+    next();
+  } catch (err) {
+    const errorMessage = 'Please try logging in again';
+    res.status(401).json({ error: errorMessage });
+  }
+};
+
+// Model
+const User = mongoose.model('User', userSchema);
+
+// PORT=9000 npm start
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
 
-// Start defining your routes here
+// Simple usage
+// This is how we have used CORS in previous project, which is a Simple Usage (Enable All CORS Requests).
+// As default the CORS allow all domain to access the server.
+// Access-Control-Allow-Origin: * is shown in the network tab.  The * = all domain has access.
+
+app.use(cors());
+app.use(bodyParser.json());
+
+// GET - list of all endpoints
+const endPointList = require('express-list-endpoints');
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  if (!res) {
+    res
+      .status(404)
+      .send({ error: 'Oops! Something goes wrong. Try again later!' });
+  }
+  res.send(endPointList(app));
+});
+
+// Configuring CORS
+// To limit access to a specific domain, below is an example how to write that in code.
+// We only allow http://localhost:3000 as an origin.
+// app.use(cors());
+// app.use(bodyParser.json());
+
+// var corsOptions = {
+//   origin: 'http://localhost:3000',
+//   optionsSuccessStatus: 200 // some legacy browsers (IE11, various SmartTVs) choke on 204
+// };
+
+// app.get('/', cors(corsOptions), (req, res) => {
+//   res.json('Hello hello');
+// });
+
+// Const's for error messages instead of text in error handling
+const SERVICE_UNAVAILABLE = 'Service unavailable.';
```

Comment:
For consistency we would recommend using either variables or displaying the message as a string in the json.

Reply:
Removed the const and added the string into the json.